lib/fluent/plugin_helper/server.rb

```diff
@@ -0,0 +1,375 @@
+
+      def server_attach(title, port, bind, shared, server)
+        @_servers << ServerInfo.new(title, port, bind, shared, server)
+        event_loop_attach(server)
+      end
+
+      def server_create_for_tcp_connection(shared, bind, port, resolve_name, linger_timeout, backlog, &block)
+        sock = server_create_tcp_socket(shared, bind, port)
+        server = Coolio::TCPServer.new(sock, nil, TCPServerHandler, resolve_name, linger_timeout, @log, @under_plugin_development, block)
+        server.listen(backlog) if backlog
+        server
+      end
+
+      def initialize
+        super
+        @_servers = []
+      end
+
+      def stop
+        super
+      end
+
+      def shutdown
+        super
+      end
+
+      def close
+        @_servers.each do |si|
+          si.server.close rescue nil
+        end
+
+        super
+      end
+
+      def terminate
+        @_servers = []
+        super
+      end
+
+      def server_certopts_validate!(certopts)
+        raise "not implemented yet"
+      end
+
+      def server_socket_manager_client
+        socket_manager_path = ENV['SERVERENGINE_SOCKETMANAGER_PATH']
+        if Fluent.windows?
+          socket_manager_path = socket_manager_path.to_i
+        end
+        ServerEngine::SocketManager::Client.new(socket_manager_path)
+      end
+
+      def server_create_tcp_socket(shared, bind, port)
+        sock = if shared
+                 server_socket_manager_client.listen_tcp(bind, port)
+               else
+                 TCPServer.new(bind, port) # this method call can create sockets for AF_INET6
+               end
+        sock.fcntl(Fcntl::F_SETFD, Fcntl::FD_CLOEXEC) # close-on-exec
+        sock.fcntl(Fcntl::F_SETFL, Fcntl::O_NONBLOCK) # nonblock
+        sock
+      end
+
+      def server_create_udp_socket(shared, bind, port)
+        sock = if shared
+                 server_socket_manager_client.listen_udp(bind, port)
+               else
+                 family = IPAddr.new(IPSocket.getaddress(bind)).ipv4? ? ::Socket::AF_INET : ::Socket::AF_INET6
+                 usock = UDPSocket.new(family)
+                 usock.bind(bind, port)
+                 usock
+               end
+        sock.fcntl(Fcntl::F_SETFD, Fcntl::FD_CLOEXEC) # close-on-exec
+        sock.fcntl(Fcntl::F_SETFL, Fcntl::O_NONBLOCK) # nonblock
+        sock
+      end
+
+      def server_create_tls_socket(shared, bind, port)
+        raise "not implemented yet"
+      end
+
+      class UDPServerHandler < Coolio::IO
+        class SocketForCallback < ::UDPSocket
+          def remote_addr; peeraddr[3]; end
+          def remote_host; peeraddr[2]; end
+          def remote_port; peeraddr[1]; end
+
+          def write(data)
+            send(data, 0) # flag - nothing specified
+          end
+        end
+
+        def initialize(sock, resolve_name, max_bytes, flags, log, under_plugin_development, &callback)
+          raise ArgumentError, "socket is a UDPSocket" unless sock.is_a?(UDPSocket)
```

Comment:
`socket must be a UDPSocket: socket = #{sock}` or `socket is not a UDPSocket: socket = #{sock}`  is better.